### ISSUE TITLE
Python 2 is being deprecated

### ISSUE THREAD

**Issue by ElonSatoshi** [cmdln/flashbake#36]
Might be a good idea to make this software work with python 3

**Comment by brainwane** [comment]
In March (prior to the 0.28 release), because this software depends on Python 2 and seemed unmaintained, [Debian removed it from unstable](https://tracker.debian.org/news/1107949/removed-0271-01-from-unstable/). @cmdln would you be interested in other contributors helping port this software to Python 3?

**Comment by brainwane** [comment]
@abhidg heads-up about how [Debian's tracker](https://tracker.debian.org/pkg/flashbake) doesn't know about this year's release.

**Comment by abhidg** [comment]
@brainwane thanks, the tracker hasn't been updated to use github I think, if it's ported to Python 3, I will take a look at the packaging, it's been a while :)

**Comment by cmdln** [comment]
I am entirely open to anyone who wants to port Flashbake to Python 3. I haven't done any Python coding in years and am happy to grant permissions for anyone else willing to do this work.

**Comment by brainwane** [comment]
@cmdln ok, great, I'm gonna quickly see how big a job that looks like. Anything I should know about which branch to work off (in case I should be looking at something other than `master`) or any other development-process gotchas?

**Comment by cmdln** [comment]
No other branches and it has been so long, I don't recall any particular gotchas.

**Comment by brainwane** [comment]
I took about 30-45 minutes to familiarize myself with stuff and saw that it won't be a piece of cake for me (however, it's been years since I converted code from Python 2 to Python 3, so it's probably a lot faster for someone else). I created a work-in-progress pull request #39 -- I am not currently a user of flashbake and probably won't be for several weeks* so I don't know when or if I'll come back to it. Happy to add other users as collaborators on that fork so you can make more progress on the PR if anyone's interested. Or you can make your own PR of course!


`*` During my writing this month, I want to track daily wordcount, and it would be a little annoying to try to do that with flashbake's model of committing multiple times per hour; right now I'm instead committing once per day so I can use [this approach](https://stackoverflow.com/a/53719737) to count my words written that day.

**Comment by ianpaul** [comment]
Whoops, sorry I didn't check the issues here before I started working. Looks like we've had some slight duplication of effort, but I see this hasn't gone very far. 

I've got an active branch here: https://github.com/ianpaul/flashbake/tree/flashbake3. 

**I have flashbake working in Python 3 with two major hiccups:**

1. Dry runs are staging files for commit, which I don't believe is the expected behavior. Am I correct @cmdln ?
2. Some plugins aren't being recognized such as Weather. 

Anyone who wants to help by adding fixes would be most welcome. 

The biggest stumbling block I hit so far had to do with the way Python 3 differentiates between byte and string types. Once I figured out where to place `.decode('utf-8')` it was pretty easy to get things moving. 

Anyway, I'll be working on this on the weekends and some weekday evenings if anyone wants to help.
